tools/find-inactive-collaborators.js

```diff
@@ -0,0 +1,77 @@
+'use strict';
+
+// Identify inactive collaborators. "Inactive" is not quite right, as the things
+// this checks for are not the entirety of collaborator activities. Still, it is
+// a pretty good proxy. Feel free to suggest or implement further metrics.
+
+const SINCE = process.argv[2] || '6 months ago';
+
+const childProcess = require('child_process');
+const fs = require('fs');
+const path = require('path');
+
+// Retrieve all commit authors during the time period.
+const authors = childProcess.execSync(
+  `git shortlog -n -s --since="${SINCE}"`,
+  { cwd: path.resolve(__dirname, '..'),
+    encoding: 'utf8',
+    stdio: ['inherit', 'pipe', 'pipe'] }
+).split('\n')
```

Comment:
Wouldn't `readline` be more suited for this use case? It would allow to spawn the `git` processes concurrently, but maybe that's something you wanted to avoid?

Reply:
I thought `git` processes couldn't run concurrently because they'd get errors from each other's lock files, but that's wrong because `git log` and `git shortlog` are read-only operations so there is no lock file. 

How would you do this in a way that minimizes the added code complexity? Promisify `exec` and then wrap them in a  `Promise.all()`?

Reply:
Yes, it would make sense to use ESM to get TLA:

```js
async function exec(cmd, mapFn) {
  const childProcess = cp.spawn('/bin/sh', ['-c', cmd], {
    cwd: new URL('..', import.meta.url),
    encoding: 'utf8',
    stdio: ['inherit', 'pipe', 'inherit'],
  });
  const lines = readline.createInterface({
    input: childProcess.stdout,
  });
  const errorHandler = new Promise((_, reject) => childProcess.on('error', reject));
  const returnedSet = new Set(); // using a set so no need to pipe to `sort | uniq`
  await Promise.race([errorHandler, Promise.resolve()]);
  for await (const line of lines) {
    await Promise.race([errorHandler, Promise.resolve()]);
    const val = mapFn(line);
    if (val) {
      returnedSet.add(val);
    }
  }
  return Promise.race([errorHandler, Promise.resolve(returnedSet)]);
}
```

Reply:
I was about to comment that this won't work on Windows but then realized neither will the code I already have. 🙃

Reply:
Awaiting the three promises in turn takes less time than running them simultaneously via `Promise.all()`. 🤔

<details>
Await in turn version:

```js
#!/usr/bin/env node

// Identify inactive collaborators. "Inactive" is not quite right, as the things
// this checks for are not the entirety of collaborator activities. Still, it is
// a pretty good proxy. Feel free to suggest or implement further metrics.

import cp from 'node:child_process';
import fs from 'node:fs';
import readline from 'node:readline';

const SINCE = process.argv[2] || '6 months ago';

async function runGitCommand(cmd, mapFn) {
  const childProcess = cp.spawn('/bin/sh', ['-c', cmd], {
    cwd: new URL('..', import.meta.url),
    encoding: 'utf8',
    stdio: ['inherit', 'pipe', 'inherit'],
  });
  const lines = readline.createInterface({
    input: childProcess.stdout,
  });
  const errorHandler = new Promise(
    (_, reject) => childProcess.on('error', reject)
  );
  const returnedSet = new Set();
  await Promise.race([errorHandler, Promise.resolve()]);
  for await (const line of lines) {
    await Promise.race([errorHandler, Promise.resolve()]);
    const val = mapFn(line);
    if (val) {
      returnedSet.add(val);
    }
  }
  return Promise.race([errorHandler, Promise.resolve(returnedSet)]);
}

// Retrieve all commit authors during the time period.
const authors = await runGitCommand(
  `git shortlog -n -s --since="${SINCE}"`,
  (line) => line.trim().split('\t', 2)[1]
);

// Retrieve all commit landers during the time period.
const landers = await runGitCommand(
  `git shortlog -n -s -c --since="${SINCE}"`,
  (line) => line.trim().split('\t', 2)[1]
);

// Retrieve all approving reviewers of landed commits during the time period.
const approvingReviewers = await runGitCommand(
  `git log --since="${SINCE}" | egrep "^    Reviewed-By: "`,
  (line) => /^    Reviewed-By: ([^<]+)/.exec(line)[1].trim()
);

// Retrieve list of current collaborators from README.md.
const readmeText = fs.readFileSync(
  new URL('../README.md', import.meta.url),
  'utf8'
);
let processingCollaborators = false;
const collaborators = readmeText
  .split('\n')
  .filter((line) => {
    const isCollaborator = processingCollaborators && line.length;
    if (line === '### Collaborators') {
      processingCollaborators = true;
    }
    if (line === '### Collaborator emeriti') {
      processingCollaborators = false;
    }
    return line.startsWith('**') && isCollaborator;
  })
  .map((line) => line.split('**')[1].trim());

console.log(`${authors.size.toLocaleString()} authors have made commits since ${SINCE}.`);
console.log(`${landers.size.toLocaleString()} landers have landed commits since ${SINCE}.`);
console.log(`${approvingReviewers.size.toLocaleString()} reviewers have approved landed commits since ${SINCE}.`);
console.log(`${collaborators.length.toLocaleString()} collaborators currently in the project.`);

const inactive = collaborators.filter((collaborator) =>
  !authors.has(collaborator) &&
  !landers.has(collaborator) &&
  !approvingReviewers.has(collaborator)
);

if (inactive.length) {
  console.log('\nInactive collaborators:');
  console.log(inactive.join('\n'));
}
```

Promise.all() version:

```js
#!/usr/bin/env node

// Identify inactive collaborators. "Inactive" is not quite right, as the things
// this checks for are not the entirety of collaborator activities. Still, it is
// a pretty good proxy. Feel free to suggest or implement further metrics.

import cp from 'node:child_process';
import fs from 'node:fs';
import readline from 'node:readline';

const SINCE = process.argv[2] || '6 months ago';

async function runGitCommand(cmd, mapFn) {
  const childProcess = cp.spawn('/bin/sh', ['-c', cmd], {
    cwd: new URL('..', import.meta.url),
    encoding: 'utf8',
    stdio: ['inherit', 'pipe', 'inherit'],
  });
  const lines = readline.createInterface({
    input: childProcess.stdout,
  });
  const errorHandler = new Promise(
    (_, reject) => childProcess.on('error', reject)
  );
  const returnedSet = new Set();
  await Promise.race([errorHandler, Promise.resolve()]);
  for await (const line of lines) {
    await Promise.race([errorHandler, Promise.resolve()]);
    const val = mapFn(line);
    if (val) {
      returnedSet.add(val);
    }
  }
  return Promise.race([errorHandler, Promise.resolve(returnedSet)]);
}

const [authors, landers, approvingReviewers] = await Promise.all([
  runGitCommand(
    `git shortlog -n -s --since="${SINCE}"`,
    (line) => line.trim().split('\t', 2)[1]
  ),
  runGitCommand(
    `git shortlog -n -s -c --since="${SINCE}"`,
    (line) => line.trim().split('\t', 2)[1]
  ),
  runGitCommand(
    `git log --since="${SINCE}" | egrep "^    Reviewed-By: "`,
    (line) => /^    Reviewed-By: ([^<]+)/.exec(line)[1].trim()
  ),
]);

// Retrieve list of current collaborators from README.md.
const readmeText = fs.readFileSync(
  new URL('../README.md', import.meta.url),
  'utf8'
);
let processingCollaborators = false;
const collaborators = readmeText
  .split('\n')
  .filter((line) => {
    const isCollaborator = processingCollaborators && line.length;
    if (line === '### Collaborators') {
      processingCollaborators = true;
    }
    if (line === '### Collaborator emeriti') {
      processingCollaborators = false;
    }
    return line.startsWith('**') && isCollaborator;
  })
  .map((line) => line.split('**')[1].trim());

console.log(`${authors.size.toLocaleString()} authors have made commits since ${SINCE}.`);
console.log(`${landers.size.toLocaleString()} landers have landed commits since ${SINCE}.`);
console.log(`${approvingReviewers.size.toLocaleString()} reviewers have approved landed commits since ${SINCE}.`);
console.log(`${collaborators.length.toLocaleString()} collaborators currently in the project.`);

const inactive = collaborators.filter((collaborator) =>
  !authors.has(collaborator) &&
  !landers.has(collaborator) &&
  !approvingReviewers.has(collaborator)
);

if (inactive.length) {
  console.log('\nInactive collaborators:');
  console.log(inactive.join('\n'));
}
```

Test command:

```
time node tools/find-inactive-collaborators.mjs '20 years ago
```

Node version: 16.4.2
OS: macOS Big Sur, 8-year-old MacBook Pro laptop

Results:  Promise.all() version takes ~3.15 seconds. Await-in-turn version takes ~2.55 seconds.

</details>

Reply:
Updated to use async/await but did not use Promise.all() because performance is worse (although by only a quarter of a second or whatever--but the code is harder to read/understand too, so if we're not getting performance out of it, then the trade-off isn't worth it). 

The current async/await version outperforms the previous execSync() version (which in turn outperforms the Promise.all() version).

I'm going to mark this conversation as resolved but if you or anyone else wants to figure out what's up with Promise.all(), by all means, please do!